damus/Util/Zap.swift

```diff
@@ -0,0 +1,307 @@
+//
+//  Zap.swift
+//  damus
+//
+//  Created by William Casarin on 2023-01-15.
+//
+
+import Foundation
+
+enum ZapSource {
+    case author(String)
+    // TODO: anonymous
+    //case anonymous
+}
+
+public struct NoteZapTarget: Equatable {
+    public let note_id: String
+    public let author: String
+}
+
+public enum ZapTarget: Equatable {
+    case profile(String)
+    case note(NoteZapTarget)
+    
+    public static func note(id: String, author: String) -> ZapTarget {
+        return .note(NoteZapTarget(note_id: id, author: author))
+    }
+    
+    var pubkey: String {
+        switch self {
+        case .profile(let pk):
+            return pk
+        case .note(let note_target):
+            return note_target.author
+        }
+    }
+    
+    var id: String {
+        switch self {
+        case .note(let note_target):
+            return note_target.note_id
+        case .profile(let pk):
+            return pk
+        }
+    }
+}
+
+struct ZapRequest {
+    let ev: NostrEvent
+}
+
+struct Zap {
+    public let event: NostrEvent
+    public let invoice: ZapInvoice
+    public let zapper: String /// zap authorizer
+    public let target: ZapTarget
+    public let request: ZapRequest
+    
+    public static func from_zap_event(zap_ev: NostrEvent, zapper: String) -> Zap? {
+        /// Make sure that we only create a zap event if it is authorized by the profile or event
+        guard zapper == zap_ev.pubkey else {
+            return nil
+        }
+        guard let bolt11_str = event_tag(zap_ev, name: "bolt11") else {
+            return nil
+        }
+        guard let bolt11 = decode_bolt11(bolt11_str) else {
+            return nil
+        }
+        /// Any amount invoices are not allowed
+        guard let zap_invoice = invoice_to_zap_invoice(bolt11) else {
+            return nil
+        }
+        guard let preimage = event_tag(zap_ev, name: "preimage") else {
```

Comment:
I think the preimage is not very useful, it also prevents some people supporting this because they may not always have access to the preimage, that is the case with strike.army also

Reply:
I can remove this check, you're mostly trusting the zapper anyways to determine if something is paid.

Reply:
I'm curious why the node sending the zap wouldn't have access to the preimage, usually it would be reacting to a payment when it sends the zap note...